flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/NonWindowJoin.scala

```diff
@@ -0,0 +1,271 @@
+
+    if (stateCleaningEnabled && leftTimer.value == timestamp) {
+      expireOutTimeRowForLeft(
+        timestamp,
+        leftState,
+        leftTimer,
+        ctx
+      )
+    }
+
+    if (stateCleaningEnabled && rightTimer.value == timestamp) {
+      expireOutTimeRowForRight(
+        timestamp,
+        rightState,
+        rightTimer,
+        ctx
+      )
+    }
+  }
+
+  /**
+    * Removes records which are expired from left state.
+    */
+  def expireOutTimeRowForLeft(curTime: Long,
+      rowMapState: MapState[Row, JTuple2[Int, Long]],
+      timerState: ValueState[Long],
+      ctx: CoProcessFunction[CRow, CRow, CRow]#OnTimerContext): Unit = {
+
+    expireOutTimeRow(curTime, rowMapState, timerState, ctx)
+  }
+
+  /**
+    * Removes records which are expired from right state.
+    */
+  def expireOutTimeRowForRight(curTime: Long,
+      rowMapState: MapState[Row, JTuple2[Int, Long]],
+      timerState: ValueState[Long],
+      ctx: CoProcessFunction[CRow, CRow, CRow]#OnTimerContext): Unit = {
+
+    expireOutTimeRow(curTime, rowMapState, timerState, ctx)
+  }
+
+
+  def getNewExpiredTime(
+      curProcessTime: Long,
+      oldExpiredTime: Long): Long = {
+
+    if (stateCleaningEnabled && curProcessTime + minRetentionTime > oldExpiredTime) {
+      curProcessTime + maxRetentionTime
+    } else {
+      oldExpiredTime
+    }
+  }
+
+  /**
+    * Removes records which are expired from the state. Registers a new timer if the state still
+    * holds records after the clean-up.
+    */
+  def expireOutTimeRow(
+      curTime: Long,
+      rowMapState: MapState[Row, JTuple2[Int, Long]],
+      timerState: ValueState[Long],
+      ctx: CoProcessFunction[CRow, CRow, CRow]#OnTimerContext): Unit = {
+
+    val rowMapIter = rowMapState.iterator()
+    var validTimestamp: Boolean = false
+
+    while (rowMapIter.hasNext) {
+      val mapEntry = rowMapIter.next()
+      val recordExpiredTime = mapEntry.getValue.f1
+      if (recordExpiredTime <= curTime) {
+        rowMapIter.remove()
+      } else {
+        // we found a timestamp that is still valid
+        validTimestamp = true
+      }
+    }
+
+    // If the state has non-expired timestamps, register a new timer.
+    // Otherwise clean the complete state for this input.
+    if (validTimestamp) {
+      val cleanupTime = curTime + maxRetentionTime
+      ctx.timerService.registerProcessingTimeTimer(cleanupTime)
+      timerState.update(cleanupTime)
+    } else {
+      timerState.clear()
+      rowMapState.clear()
+    }
+  }
+
+  /**
+    * Connect left row and null right row, then collect. The result is NULL from the right side,
+    * if there is no match.
+    *
+    * @param leftRow    The row from left side.
+    * @param defaultRow The result row used for output, right side fields will all be null.
+    * @param out        The collector for returning result values.
+    */
+  def collectWithNullRight(leftRow: Row, defaultRow: Row, out: Collector[Row]): Unit = {
```

Comment:
I was trying to reduce the number of `if else` as much as possible which is inefficient. What do you think?

Reply:
I was thinking about somthing like 
```
def collectWithDefault(mainRow: Row, defaultRow: Row, out: Collector[Row]): Unit = {
   //...
}
```
This way `mainRow` will be `leftRow` when LEFT JOIN and `rightRow` when RIGTH JOIN. It might be usable in FULL OUTER JOIN as well.